great_expectations/core/util.py

```diff
@@ -782,150 +779,63 @@ def sniff_s3_compression(s3_url: S3Url) -> Union[str, None]:
-
-    return spark_session
-
-
-# noinspection PyPep8Naming
 def get_or_create_spark_session(
-    spark_config: Optional[Dict[str, str]] = None,
-) -> pyspark.SparkSession | None:
+    spark_config: Optional[dict[str, str]] = None,
+) -> pyspark.SparkSession:
     """Obtains Spark session if it already exists; otherwise creates Spark session and returns it to caller.
 
-    Due to the uniqueness of SparkContext per JVM, it is impossible to change SparkSession configuration dynamically.
-    Attempts to circumvent this constraint cause "ValueError: Cannot run multiple SparkContexts at once" to be thrown.
-    Hence, SparkSession with SparkConf acceptable for all tests must be established at "pytest" collection time.
-    This is preferred to calling "return SparkSession.builder.getOrCreate()", which will result in the setting
-    ("spark.app.name", "pyspark-shell") remaining in SparkConf statically for the entire duration of the "pytest" run.
-
     Args:
         spark_config: Dictionary containing Spark configuration (string-valued keys mapped to string-valued properties).
 
     Returns:
-
+        SparkSession
     """
-    spark_session: Optional[pyspark.SparkSession]
-    try:
-        if spark_config is None:
-            spark_config = {}
-        else:
-            spark_config = copy.deepcopy(spark_config)
+    spark_config = spark_config or {}
 
+    try:
         builder = pyspark.SparkSession.builder
 
-        app_name: Optional[str] = spark_config.get("spark.app.name")
-        if app_name:
-            builder.appName(app_name)
-
+        # user could get in trouble here if they try to set config options that are not allowed in their runtime
         for k, v in spark_config.items():
             if k != "spark.app.name":
                 builder.config(k, v)
 
-        spark_session = builder.getOrCreate()
-        # noinspection PyProtectedMember,PyUnresolvedReferences
-        if spark_session.sparkContext._jsc.sc().isStopped():
-            raise ValueError("SparkContext stopped unexpectedly.")
+        app_name: str | None = spark_config.get("spark.app.name")
+        if app_name:
+            builder.appName(app_name)
+
+        spark_session: pyspark.SparkSession = builder.getOrCreate()
 
-    except AttributeError:
+        if in_databricks(spark_session=spark_session):
+            if app_name:
+                warnings.warn(
+                    "Passing spark.app.name to spark_config has no effect in a Databricks environment.",
+                    category=RuntimeWarning,
+                )
+        else:
+            # in a local pyspark-shell the context config cannot be updated
+            # unless you stop the Spark context and re-recreate it
+            retrieved_spark_config: pyspark.SparkConf = (
+                spark_session.sparkContext.getConf()
+            )
+            stopped = False
+            for key, value in spark_config.items():
+                if retrieved_spark_config.get(key) != value:
```

Comment:
We check whether the current spark config differs from the input spark config, and I see the current spark session is stopped then restarted if there are differing config values. Are the input `spark_config` values (the one supplied `get_or_create_spark_session`) supposed get set for the restarted session? If so, where does this happen?

Reply:
The input `spark_config` is passed to `builder.config` already, but for local environments it is ignored if there is an existing Spark context. We don't need to pass the config to `builder` again, we just need to stop and restart with the same `builder`.